[ci-metrics] add docs describing the metrics collected (#78363)
elastic/kibana: docs/developer/contributing/development-ci-metrics.asciidoc

[[ci-metrics]]
== CI Metrics

> In addition to run our tests CI collects metrics about Kibana. These metrics are sent to an external service which catalogs and compares them to the metrics from other branches so that PR authors can better quantify the impact their changes have.

Reviewer: ```suggestion
In addition to running our tests, CI collects metrics about the Kibana build. These metrics are sent to an external service to track changes over time, and to provide PR authors insights into the impact of their changes.
```

[[ci-metric-types]]
=== Metric types

[[ci-metric-types-bundle-size-metrics]]
==== Bundle size

These metrics help contributors know how they are impacting the size of the bundles Kibana creates, and help make sure that Kibana loads as fast as possible.

[[ci-metric-page-load-bundle-size]] `page load bundle size` ::

> The size of the entry file produced for each bundle/plugin. This file is always loaded on every page load, so it should be as small as possible. To reduce this metric you can put any code that isn't necessary on every page load behind an https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Dynamic_Imports[`async import()`].

Reviewer: Should we link to https://github.com/elastic/kibana/blob/master/src/core/MIGRATION.md#keep-kibana-fast, or move that particular part to docs? cc @joshdover

Author: I like the idea of linking there

Author: Oh, didn't realize those weren't already in the docs. We should probably move that to the docs and link to there then. Linking to a section of the MIGRATION.md file feels like it's going to break pretty quickly.

Reviewer: I don't think we should block this PR on that change, but we should add the link once those docs are moved.

+

> Code that is shared statically with other plugins, including anything that's exported from the `public/index.ts` file of a plugin and any code referenced by the `extraPublicDirs` manifest property, is always exported from the entry file meaning it will contribute to the `page load bundle size` for the plugin which is sharing the code.

Reviewer: A little hard to read, I suggest: 

```suggestion
Code that is shared statically with other plugins will contribute to the `page load bundle size` of that plugin. This includes exports from the `public/index.ts` file and any file referenced by the `extraPublicDirs` manifest property. 
```

[[ci-metric-async-chunks-size]] `async chunks size` ::
An "async chunk" is created for the files imported by each https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Dynamic_Imports[`async import()`] statement. This metric tracks the sum size of these chunks, in bytes, broken down by plugin/bundle id. You can think of this as the amount of code users will have to download if they access all the components/applications within a bundle.

[[ci-metric-misc-asset-size]] `miscellaneous assets size` ::
A "miscellaneous asset" is anything that isn't an async chunk or entry chunk, often images. This metric tracks the sum size of these assets, in bytes, broken down by plugin/bundle id.

[[ci-metric-bundle-module-count]] `@kbn/optimizer bundle module count` ::
The number of separate modules included in each bundle/plugin. This is the best indicator we have for how long a specific bundle will take to be built by the `@kbn/optimizer`, so we report it to help people know when they've imported a module which might include a surprising number of sub-modules.

[[ci-metric-types-distributable-size]]
==== Distributable size

> The size of the Kibana distributable is an important metric to keep an eye not only so it's fast for people to download and use Kibana, but also because the number of files in our distributable directly impacts the speed of extracting a Kibana distributable.

Reviewer: ```suggestion
The size of the Kibana distributable is an essential metric as it not only contributes to the time it takes to download, but it also impacts time it takes to extract the archive once downloaded.
```
